### ISSUE TITLE
Update() requires either a single JavaScript object or an alternating list of field/value pairs that can be followed by an optional precondition. Value for argument "preconditionOrValues" is not a valid precondition. "exists" is not an allowed precondition

### ISSUE THREAD

**Issue by alexandr2levin** [googleapis/nodejs-firestore#1846]
#### Environment details

  - OS: macOS 13.1
  - Node.js version: v16.19.0
  - npm version: 8.19.3
  - `@google-cloud/firestore` version: 6.5.0 (as part of firebase-admin@11.7.0)

#### Steps to reproduce

1. Try to update a set of firestore documents but keep the batch from failing if some of them are missing by providing the `{ exists: false }` precondition parameter:

```
    const productBatch = getFirestore().batch();
    for(const productReference of productReferences) {
      productBatch.update(productReference, {
        [nameof<ProductDocument>((o) => o.stock)]: FieldValue.increment(-1),
      }, { exists: false });
    }
    await productBatch.commit();
```

2. This is the error:
```
Unhandled error Error: Update() requires either a single JavaScript object or an alternating list of field/value pairs that can be followed by an optional precondition. Value for argument \"preconditionOrValues\" is not a valid precondition. \"exists\" is not an allowed precondition.
    at WriteBatch.update (/Users/wok/Projects/selftable_cf_4_home/functions/node_modules/@google-cloud/firestore/build/src/write-batch.js:392:23)
    at /Users/wok/Projects/selftable_cf_4_home/functions/lib/src/features/order.js:217:26
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async /Users/wok/Projects/selftable_cf_4_home/functions/node_modules/firebase-functions/lib/common/providers/https.js:438:26
    at async runFunction (/Users/wok/Projects/selftable_cf_4_home/functions/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:506:9)
    at async runHTTPS (/Users/wok/Projects/selftable_cf_4_home/functions/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:531:5)
    at async /Users/wok/Projects/selftable_cf_4_home/functions/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:690:21
```

I saw another issue stating a similar problem (https://github.com/googleapis/nodejs-firestore/issues/1657) but that case is a bit different — `{ exists: true }` was passed which is the default behavior of the .update() function so there's no actual need to pass this parameter. In my case, i want to change the default behavior by passing `{ exists: false }`.

The update function declaration clearly states that it expects me to pass the `Precondition` type:

<img width="670" alt="image" src="https://user-images.githubusercontent.com/5736157/235898326-2761c50c-deca-4cec-9326-35f86efd6578.png">

Which looks like this and expects the `exists` parameter to be passed:

<img width="674" alt="image" src="https://user-images.githubusercontent.com/5736157/235898188-702f491a-2963-4de0-aefc-5329fb22d00b.png">

If it's not allowed to pass the `exists` parameter at all, maybe it would be better to declare the UpdatePrecondition interface for this function specifically.

**Comment by alexandr2levin** [comment]
The workaround would be to use a transaction:
```
    await getFirestore().runTransaction(async (transaction) => {
      const emptyProductSnapshots = (await transaction.getAll(...productReferences, { fieldMask: [] }));
      for(const emptyProductSnapshot of emptyProductSnapshots) {
        transaction.update(emptyProductSnapshot.ref, {
          [nameof<ProductDocument>((o) => o.stock)]: FieldValue.increment(1),
        });
      }
    });
```
But it leads to potentially a lot of unnecessary read operations.

**Comment by dconeybe** [comment]
Hi @alexandr2levin. Thank you for pointing out your issue with preconditions.

Unfortunately, "update" operations do _not_ support `{exists: false}` as a precondition. I agree that it is misleading that the public APIs allow callers to specify this unsupported precondition and your suggestion of breaking apart `Precondition` into something like `UpdatePrecondition` and `DeletePrecondition` makes sense to me. I'll bring this up with my team and see what we think about your suggestion (Googlers see b/280644695).

Even if we _did_ allow `{exists: false}` as a precondition it doesn't sound like it would suit your use case. That precondition means "fail the operation if the document exists". This doesn't really make sense for an "update" operation and it does not provide the "create the document if it doesn't exist" semantics.

Your workaround of using a transaction looks good and optimal. Alternately, have you considered using `set()` with `merge=true`? It sounds like it would handle your specific use case. Here is the documentation for `SetOptions` for more details.: https://github.com/googleapis/nodejs-firestore/blob/fda12427184df2eff0b4de262ec8a7c5fa430702/types/firestore.d.ts#L1130-L1152

**Comment by alexandr2levin** [comment]
Hi @dconeybe! Thank you for your reply.

> Even if we did allow {exists: false} as a precondition it doesn't sound like it would suit your use case. That precondition means "fail the operation if the document exists". This doesn't really make sense for an "update" operation and it does not provide the "create the document if it doesn't exist" semantics.

> Your workaround of using a transaction looks good and optimal. Alternately, have you considered using set() with merge=true? It sounds like it would handle your specific use case. Here is the documentation for SetOptions for more details.:

Sorry, my workaround was misleading, here's the fixed version:
```
    await getFirestore().runTransaction(async (transaction) => {
      const emptyProductSnapshots = (await transaction.getAll(...productReferences, { fieldMask: [] }));
      for(const emptyProductSnapshot of emptyProductSnapshots) {
        if(emptyProductSnapshot.exists) {
          transaction.update(emptyProductSnapshot.ref, {
            [nameof<ProductDocument>((o) => o.stock)]: FieldValue.increment(-1),
          });
        }
      }
    });
```

What i'm trying to achieve with `.update({ ... }, { exists: false })` is to update a document but if it doesn't exist — do not throw an error and just finish successfully. I suppose it can be helpful for batches updating a set of documents. This way all existing documents will be successfully updated, and unexisting ones will be just ignored. Without `{ exists: false }`, i expect the batch to fail without writing anything in case some of the documents don't exist.

`.set(...)` doesn't suit my case because it will create nonexisting documents instead of ignoring them.

**Comment by dconeybe** [comment]
I see. In this case your transaction solution is the only solution I can think of with your desired semantics. I'm going to mark this issue as "closed" since it sounds like you have a workaround for this limitation in Firestore's API. Feel free to comment again to continue the discussion.

**Comment by giacomoran** [comment]
I got bit by this, in the context of the `BulkWriter`.

Given that:
- `update` fails if the doc does not exists
- `Precondition` type in `update` signature includes `exists: boolean`
- nowhere is it stated explicitly what happens if `Precondition` is not satisfied

I assumed that updates not satisfying the `Precondition` would be silently skipped.

---

As an aside `lastUpdateTime` documentation could also be improved, only from the example I was able to understand that the condition is a `==` instead of `<=`.
